cmd/dep/ensure.go

```diff
@@ -731,18 +737,46 @@ func checkErrors(m map[string]pkgtree.PackageOrErr) error {
 			case *build.NoGoError:
 				noGoErrors++
 			default:
-				buildErrors = append(buildErrors, poe.Err.Error())
+				pkgtreeErrors.errs = append(pkgtreeErrors.errs, poe.Err)
 			}
 		}
 	}
 
+	// If pkgtree was empty or all dirs lacked any Go code, return an error.
 	if len(m) == 0 || len(m) == noGoErrors {
-		return errors.New("all dirs lacked any go code")
+		return errors.New("no dirs contained any Go code")
+	}
+
+	// If all dirs contained build errors, return an error.
+	if len(m) == len(pkgtreeErrors.errs) {
+		return errors.New("all dirs contained build errors")
+	}
+
+	// If all directories either had no Go files or caused a build error, return an error.
+	if len(m) == len(pkgtreeErrors.errs)+noGoErrors {
+		return pkgtreeErrors
 	}
 
-	if len(buildErrors) > 0 {
-		return errors.Errorf("Found %d errors:\n\n%s", len(buildErrors), strings.Join(buildErrors, "\n"))
+	// If m contained some errors, return a warning with these errors.
+	if len(pkgtreeErrors.errs) > 0 {
+		pkgtreeErrors.warn = true
+		return pkgtreeErrors
```

Comment:
If this one should be interpreted as a warning, then that is a good argument for returning `(warn, err error)` from this function